src/config/compiler.rs

```diff
@@ -82,3 +84,139 @@ pub(super) fn expand_macros(config: &mut Config) -> Result<(), Vec<String>> {
         Ok(())
     }
 }
+
+/// Expand trailing `*` wildcards in input lists
+fn expand_wildcards(config: &mut Config) {
+    let candidates = config
+        .sources
+        .keys()
+        .chain(config.transforms.keys())
+        .cloned()
+        .collect::<Vec<String>>();
+
+    for (name, transform) in config.transforms.iter_mut() {
+        expand_inner(&mut transform.inputs, name, &candidates);
+    }
+
+    for (name, sink) in config.sinks.iter_mut() {
+        expand_inner(&mut sink.inputs, name, &candidates);
+    }
+}
+
+fn expand_inner(inputs: &mut Vec<String>, name: &str, candidates: &[String]) {
```

Comment:
I suggest `expand_wildcards` or `expand_names` instead.

Could this be reworked to drop the `mut` parameter and return `impl Iterator<Item = String>` (or `IntoIterator`), and then call `names.extend(expand_wildcards(name, &candidates));`? I see the non-wildcard case complicates that.

This could also generate the list of sources on the fly, without creating the intermediate vector or cloning any strings until necessary:
```rust
    config
        .sources
        .keys()
        .chain(self.transforms.key())
        .filter(|candidate| candidate.starts_with(prefix) && candidate != name)
        .map(|candidate| candidate.clone())
```

Which prompts the question of it it too could be part of `impl Config`.

Reply:
> Could this be reworked to drop the `mut` parameter and return `impl Iterator<Item = String>` (or `IntoIterator`), and then call `names.extend(expand_wildcards(name, &candidates));`?

I think the main bit that doesn't quite fit is that you need to _remove_ the wildcard itself from the list, not just add the inputs that it expands into. There are definitely some helpers that could be extracted onto `Config` or similar though.